### PR TITLE
[express-jwt] Widen ErrorCode to string, supporting autocomplete

### PR DESCRIPTION
**Pull request by womcauliff**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [The multi-tenancy example shows `"missing_secret"` as the code for the `UnauthorizedError`](https://github.com/auth0/express-jwt/blob/master/test/multitenancy.test.js)

### CODE REVIEW

**Comment by domdomegg** [comment]
From what I see in `UnauthorizedError`, any code could be potentially thrown in the `jwt.SecretCallback`.

Should this actually be widened to just `string`?

**Comment by womcauliff** [comment]
> From what I see in `UnauthorizedError`, any code could be potentially thrown in the `jwt.SecretCallback`.
> 
> Should this actually be widened to just `string`?

@domdomegg I _could_ update `ErrorCode` to simply be a `string` type.

On the other hand, I suppose the advantage of the string literal union type is that it defines the error code values that the `express-jwt` library returns _by default_, such as `invalid_token` and `revoked_token`.

There's an open issue requesting that the full list of error codes be documented https://github.com/auth0/express-jwt/issues/216 so it might be valuable to some devs to have the list.

**Comment by domdomegg** [comment]
> the string literal union type is that it defines the error code values that the `express-jwt` library returns _by default_, such as `invalid_token` and `revoked_token`.

As far as I can see in that library, `missing_secret` is not a error code that `express-jwt` returns. It isn't used in any of the code beyond the test example. It appears that there's a hook (that is used in the example) and it shows you can return arbitrary strings in the error message.

https://github.com/auth0/express-jwt/search?q=missing_secret

**Comment by womcauliff** [comment]
Fair enough, I'll widen to `string`

File: types/express-jwt/index.d.ts [hunk 1]
```diff
@@ -15,6 +15,14 @@ export = jwt;
...
     type secretType = string | Buffer;
+    /**
+     * The express-jwt library specifies the following ErrorCode values by default:
+     * - `"revoked_token"`
+     * - `"invalid_token"`
+     * - `"credentials_bad_scheme"`
+     * - `"credentials_bad_format"`
+     * - `"credentials_required"`
+     */
     type ErrorCode = string;
```
Comment (domdomegg) [inline types/express-jwt/index.d.ts:26]: If you keep the previous type and union it with string, I believe Typescript etc. will treat it just as a string, but IDEs like VS Code will suggest autocompleting these for better developer experience (while still allowing any strings)

Comment (womcauliff) [reply types/express-jwt/index.d.ts:26]: @domdomegg I added the `string` type to the union of string literals, and unfortunately, autocomplete wasn't triggered in VS Code.
```ts
    type ErrorCode =
        | 'revoked_token'
        | 'invalid_token'
        | 'credentials_bad_scheme'
        | 'credentials_bad_format'
        | 'credentials_required'
        | string;
```

I found a relevant discussion thread on the main TS repo of others describing the same issue (https://github.com/microsoft/TypeScript/issues/29729) and came across the following pattern which _would_ support autocomplete while also allowing any `string`:

```ts
    type LiteralUnion<T extends U, U = string> = T | (U & Record<never, never>);

    type ErrorCode = LiteralUnion<
        'revoked_token' | 'invalid_token' | 'credentials_bad_scheme' | 'credentials_bad_format' | 'credentials_required'
    >;
```

autocomplete:
Screen Shot 2021-12-05 at 11 31 17 AM

Thoughts?

Comment (peterblazejewicz) [reply types/express-jwt/index.d.ts:26]: it's used here, I've used it e.g. for `markdown-it-emoji`:
[types/markdown-it-emoji](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/markdown-it-emoji/index.d.ts#L25)

Comment (womcauliff) [reply types/express-jwt/index.d.ts:26]: good to know! I'll push a commit with this approach

**Review by domdomegg — APPROVED** [review]

**Review by peterblazejewicz — APPROVED** [review]

**Comment by womcauliff** [comment]
Ready to merge

**Pull request merged**